### PR TITLE
Centralizes images into agnhost (part 4)

### LINKED ISSUE #80194: Upgrade agnhost image to add build-tools
**What would you like to be added**:
Add `build-tools` package to include `dig` (+nslookup) tool in Agnhost image.

**Why is this needed**:
`dig` is important tool used in most of the network testing suite.
[Issue Source]: https://github.com/kubernetes/kubernetes/pull/79667#discussion_r303560200

/area conformance
/sig testing
/assign @bclau

### PR DESCRIPTION
**Pull request by claudiubelu**
**What type of PR is this?**

/kind feature

/sig testing
/sig windows
/area conformance

**What this PR does / why we need it**:

Centralizes the following images into ``agnhost``:

- dnsutils
- mounttest
- resource-consumer-controller
- test-webserver

Adds Windows support for ``mounttest``.

Adds CoreDNS to agnhost image, which can be used in some DNS related tests (dnsmasq is Linux-only).

Bumps ``agnhost`` version to 2.9.
Bumps ``kitten`` version to 1.1.
Bumps ``nautilus`` version to 1.1.

**Which issue(s) this PR fixes**:

Related: #76342
Fixes #80194

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: test/images/agnhost/mounttest/filePermissions.ps1
Comment (spiffxp) [inline test/images/agnhost/mounttest/filePermissions.ps1]: I lack sufficient windows expertise to really review this; can we get someone else to take a look at this file?

File: test/images/agnhost/resource-consumer/consume-cpu/consume_cpu_windows.go
Comment (spiffxp) [inline test/images/agnhost/resource-consumer/consume-cpu/consume_cpu_windows.go]: nit: this is the loop in the linux version, don't think it needs to be different here for windows
```suggestion
	for time.Since(start) < duration {
```

File: test/images/agnhost/resource-consumer/consume-cpu/consume_cpu_windows.go
Comment (spiffxp) [inline test/images/agnhost/resource-consumer/consume-cpu/consume_cpu_windows.go]: I lack sufficient windows expertise to really review this; can we get someone else to take a look at this file?

I am mostly willing to take your word on this one if the tests are happy

File: test/images/kitten/BASEIMAGE
Comment (spiffxp) [inline test/images/kitten/BASEIMAGE]: these should be 2.6, and kitten/VERSION will need a bump

File: test/images/nautilus/BASEIMAGE
Comment (spiffxp) [inline test/images/nautilus/BASEIMAGE]: these should be 2.6, and nautilus/VERSION will need a bump

**Review by spiffxp — COMMENTED** [review]
Mostly looks good, except for the two images that were based off of test-webserver

Attempting to build/push agnhost:2.6 though if you see no follow-up comment indicating I've finished, please ping me tomorrow

I would appreciate someone with windows expertise looking through some of the windows-specific files

**Comment by spiffxp** [comment]
pushed agnhost:2.6 https://gist.github.com/spiffxp/0a3c5c1c6e8b33fde786252ab9c4c765

File: test/images/agnhost/mounttest/filePermissions.ps1
Comment (bclau) [reply test/images/agnhost/mounttest/filePermissions.ps1]: FWIW, it's what it has been used in tests like ``should be consumable from pods in volume with mappings`` since about 1 year. You can see it here: https://testgrid.k8s.io/sig-windows#aks-engine-azure-windows-master

File: test/images/agnhost/resource-consumer/consume-cpu/consume_cpu_windows.go
Comment (bclau) [reply test/images/agnhost/resource-consumer/consume-cpu/consume_cpu_windows.go]: FWIW, there used to be a problem with this, which has been fixed some time ago: https://github.com/kubernetes-sigs/windows-testing/pull/20

@benmoss could give his OK on this, as he also signed off that change.

File: test/images/agnhost/resource-consumer/consume-cpu/consume_cpu_windows.go
Comment (bclau) [reply test/images/agnhost/resource-consumer/consume-cpu/consume_cpu_windows.go]: Done

File: test/images/kitten/BASEIMAGE
Comment (bclau) [reply test/images/kitten/BASEIMAGE]: oups forgot to update this.

File: test/images/nautilus/BASEIMAGE
Comment (bclau) [reply test/images/nautilus/BASEIMAGE]: Done

File: test/images/agnhost/resource-consumer/consume-cpu/consume_cpu_windows.go
Comment (benmoss) [reply test/images/agnhost/resource-consumer/consume-cpu/consume_cpu_windows.go]: Yup, this matches the work @astrieanna and I did in the old resource-consumer image.

**Comment by spiffxp** [comment]
New test has been added to the verify job since your last commit
```
Errors from staticcheck:
test/images/agnhost/resource-consumer/consume-memory/consume_memory.go:40:2: var mem is unused (U1000)

Please review the above warnings. You can test via:
  staticcheck -checks "all,-S1*,-ST1*"
```

File: test/images/agnhost/resource-consumer/consume-cpu/consume_cpu_windows.go
Comment (bclau) [reply test/images/agnhost/resource-consumer/consume-cpu/consume_cpu_windows.go]: Also, the autoscaling tests that are using the resource-consumer / resource-consumer-controller bits have been added to https://testgrid.k8s.io/sig-windows#aks-engine-azure-windows-master and they're green.

**Comment by aojea** [comment]
/cc
This is great, thanks for adding the webserver

**Comment by liggitt** [comment]
FYI, a separate update to agnhost:2.6 merged to master in https://github.com/kubernetes/kubernetes/pull/82185 (apologies, I didn't know this was in flight). This should bump to 2.7 and new agnhost images would need to be pushed

**Comment by dims** [comment]
cc @kubernetes/release-team-leads

**Comment by dims** [comment]
/assign @luxas @mkumatag @ixdy @listx

Can one of you please approve/lgtm?

**Review by johnSchnake — APPROVED** [review]
/lgtm

Can't approve so its more of a gesture until an approver comes around :)

File: test/images/agnhost/Dockerfile
Comment (aojea) [inline test/images/agnhost/Dockerfile]: This CoreDNS version has a bug that required  to move to 1.6.2 instead of using 1.5.0 https://github.com/kubernetes/kubeadm/issues/1745

File: test/images/nautilus/BASEIMAGE
Comment (aojea) [inline test/images/nautilus/BASEIMAGE]: 2.6 or 2.7?

File: test/images/agnhost/resource-consumer/consume-cpu/main.go
Comment (oomichi) [inline test/images/agnhost/resource-consumer/consume-cpu/main.go]: How about `Consumes the given amount of CPU over the given amount of time` like consume-memory?

File: test/images/agnhost/resource-consumer/consume-cpu/main.go
Comment (oomichi) [inline test/images/agnhost/resource-consumer/consume-cpu/main.go]: ditto: `Consumes the given amount of CPU over the given amount of time. This command is also used by the \"resource-consumer\" subcommand.`

**Review by oomichi — COMMENTED** [review]
Nice implementation of resource-consumer for windows, some comments inline.

File: test/images/agnhost/Dockerfile
Comment (BenTheElder) [reply test/images/agnhost/Dockerfile]: xref https://github.com/kubernetes/kubernetes/blob/master/build/dependencies.yaml

this might get weird with kube-up and kubeadm and whatnot being out of sync.

**Comment by xmudrii** [comment]
As the code freeze has started, I'm moving it to the 1.17 milestone.

/milestone v1.17

File: test/images/agnhost/README.md
Comment (bclau) [reply test/images/agnhost/README.md]: good catch!

File: test/images/agnhost/mounttest/mt_utils_windows.go
Comment (bclau) [reply test/images/agnhost/mounttest/mt_utils_windows.go]: Done

File: test/images/agnhost/resource-consumer/consume-cpu/main.go
Comment (bclau) [reply test/images/agnhost/resource-consumer/consume-cpu/main.go]: oups, forgot to update these.

File: test/images/nautilus/BASEIMAGE
Comment (bclau) [reply test/images/nautilus/BASEIMAGE]: 2.7, forgot to update this when the image version was bumped before.

**Comment by bclau** [comment]
I had to remove ``resource-consumer`` from the centralization, since we've observed some issues in the tests that consume that image. More details about this issue can be seen here: https://github.com/kubernetes/kubernetes/issues/82287

Since the ``agnhost`` Windows image is based on ``windows/servercore``, which has the problem mentioned in the issue above, we still need the image ``resource-consumer`` separated until the issue is fixed (the Windows ``resource-consumer`` is currently based on ``windows/nanoserver`` which doesn't have the same issue, and from my experiments, seem to pass the HPA tests without any problem).

As for the reason why the Windows ``agnhost`` image is using ``windows/servercore`` and not ``windows/nanoserver``: ``windows/nanoserver`` is more limited in what tools it has available, one particular pain point is the lack of powershell (which is used in its ``Dockerfile`` to install and setup things).

**Comment by BenTheElder** [comment]
> As for the reason why the Windows agnhost image is using windows/servercore and not windows/nanoserver: windows/nanoserver is more limited in what tools it has available, one particular pain point is the lack of powershell (which is used in its Dockerfile to install and setup things).

if it's only used at build time, it seems like we should look into removing that requirement.

**Comment by BenTheElder** [comment]
generally looks good to me, though I also have limited windows / powershell expertise, I'll lend an approve and leave the LGTM
/approve

**Comment by bclau** [comment]
/hold

The image version has been bumped. We should build `` gcr.io/kubernetes-e2e-test-images/agnhost:2.7`` first before we merge.

**Comment by timothysc** [comment]
cc @kubernetes/test-infra-admins  
^ Is this the right re-director?

**Comment by fejta** [comment]
Here's the process for publishing images: https://github.com/kubernetes/test-infra/blob/master/config/jobs/image-pushing/README.md

**Comment by bclau** [comment]
So, from what I understand, apparently, there are a few bits of work that will have to be done beforehand in order for the automatic image promotion to be put in place for the testing images.

First of all, we would probably need a staging repo like "gcr.io/kubernetes-staging-e2e-test-images". Do we have something like that first? And if not, can anyone create it?

Secondly, it seems that we'll have to to do a few new PRs in order to get that ready: on ``k/k``, we'll have to add  ``cloudbuild.yaml`` which will contain the steps necessary to build the images( Ihave some questions / concerns regarding this, though***), and one ``k/test-infra`` which will define image-pushing job.

*** There are a few things I am wondering about:

- when it comes to test images, we still have ~31 images in ``kubernetes/test/images``, even after all the centralization effort. Will this mean that the job will have to rebuild / publish / push / promote all the images, even if we're modifying a single image?
- there is still no validation / verification that the changes made to the images do not break anything in testgrid. Unlike most other images, any breaks here will then cause the whole merging process to stop. Ofc, this can still be mitigated by having the image updates and the image versions used by E2E tests in separate PRs, like this is. Either way, making sure there's 0% chances to break something is a bit unlikely (as we're aproaching 5K E2E tests), so there could be some follow-up PRs with some fixes. But a job using the staging repo and runnining only the ``[Conformance]`` tests would be helpful, IMO.
- hm, I think we might need a script, or some updates in ``kubernetes/test/images`` which will build / push **all** the images. Right now, you can ``make`` only one image at a time: ``make WHAT=agnhost``. We'll have to take into account that some images depend on other images: (``mounttest-user`` depends on ``mounttest`` (``mounttest-user`` is planned to be removed), ``nautilus`` and ``kitten`` depends on ``test-webserver``).

**Comment by fejta** [comment]
Will you ask @Katharine and/or #wg-k8s-infra on slack about this? They should be able to set you up with a staging repo.

**Comment by smourapina** [comment]
@bclau Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on November 18. Taking into account the extra work mentioned in your comment, is this PR still intended for 1.17?

**Comment by bclau** [comment]
I've raised these questions at the K8S infra WG meeting, you can see the minutes here: https://docs.google.com/document/d/16VBfsFMynA7tObzuZGPpw-sKDKfFc_T5W_E4IeEIaOQ/edit#

But in summary:

> 
> 
> So, from what I understand, apparently, there are a few bits of work that will have to be done beforehand in order for the automatic image promotion to be put in place for the testing images.
> 
> First of all, we would probably need a staging repo like `gcr.io/kubernetes-staging-e2e-test-images`. Do we have something like that first? And if not, can anyone create it?

I've sent a PR for this here: https://github.com/kubernetes/k8s.io/pull/400 . The PR will have to merge, then the google group will have to be created, and then the registry.

> 
> Secondly, it seems that we'll have to to do a few new PRs in order to get that ready: on `k/k`, we'll have to add `cloudbuild.yaml` which will contain the steps necessary to build the images( Ihave some questions / concerns regarding this, though***), and one `k/test-infra` which will define image-pushing job.

I've added PRs for this: https://github.com/kubernetes/kubernetes/pull/84058 and https://github.com/kubernetes/test-infra/pull/14833

> 
> *** There are a few things I am wondering about:
> 
>     * when it comes to test images, we still have ~31 images in `kubernetes/test/images`, even after all the centralization effort. Will this mean that the job will have to rebuild / publish / push / promote all the images, even if we're modifying a single image?

It seems like test-infra has a lazy-build based on directory modifications, which would solve this.

> 
>     * there is still no validation / verification that the changes made to the images do not break anything in testgrid. Unlike most other images, any breaks here will then cause the whole merging process to stop. Ofc, this can still be mitigated by having the image updates and the image versions used by E2E tests in separate PRs, like this is. Either way, making sure there's 0% chances to break something is a bit unlikely (as we're aproaching 5K E2E tests), so there could be some follow-up PRs with some fixes. But a job using the staging repo and runnining only the `[Conformance]` tests would be helpful, IMO.

Seems that this will remain as is. So, the PRs would look something like this:

- PR1: make the necessary changes to an image.
- Merge the PR.
- The Image Promoter will build and publish the image to the staging registry, from which it will be promoted to the regular registry.
- PR2: bump the image version used im ``kubernetes/test/utils/image/manifest.go`` and merge. If something breaks, revert.

> 
>     * hm, I think we might need a script, or some updates in `kubernetes/test/images` which will build / push **all** the images. Right now, you can `make` only one image at a time: `make WHAT=agnhost`. We'll have to take into account that some images depend on other images: (`mounttest-user` depends on `mounttest` (`mounttest-user` is planned to be removed), `nautilus` and `kitten` depends on `test-webserver`).

Added here: https://github.com/kubernetes/kubernetes/pull/84058

> 
> 
> @bclau Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on November 18. Taking into account the extra work mentioned in your comment, is this PR still intended for 1.17?

I've completed the items listed above. They have to be reviewed and merged. Then, we should probably be done with that and we can then merge this.

**Comment by smourapina** [comment]
Thank you for the comprehensive update @bclau. Would you then say that it is feasible to have everything merged until the code freeze date? And would it make sense to add the open PRs mentioned in your comment to the 1.17 milestone?

**Comment by bclau** [comment]
> 
> 
> Thank you for the comprehensive update @bclau. Would you then say that it is feasible to have everything merged until the code freeze date? And would it make sense to add the open PRs mentioned in your comment to the 1.17 milestone?

I would say yes. I've got a few reviews on them as well, so they're getting closer to getting merged, hopefully.

And I would say yes, it'll definitely be helpful to have them added to the milestone. It'll help the test image changes get in smoother and without having to rely on a human to build and publish the affected images.

**Comment by smourapina** [comment]
@bclau Friendly reminder: code freeze is rapidly approaching and will take place in November 14th. Is this issue still intended for 1.17? Otherwise we will need to move it to 1.18 milestone.

**Comment by bclau** [comment]
/hold cancel

Iti is the next PR in the queue for agnhost, it is approved, and it had the lgtm label at one point.

**Comment by bclau** [comment]
Hm. I'd rather let the reviewers focus on the oher, more important PRs, like the Image Promoter for v1.17. We can punt this to the next release.

**Comment by claudiubelu** [comment]
The Image Promoter bits are in place, so, merging this PR should cause the next agnhost image to be automatically built and published. We should be able to see the job here: https://testgrid.k8s.io/sig-testing-images#post-kubernetes-push-images

So, I think we can proceed with this PR.

**Comment by dims** [comment]
/lgtm

let's give this a shot!

**Pull request merged**
